### ISSUE TITLE
Huge memory usage with OSD always on and some web streams

### ISSUE THREAD

**Issue by ale-rinaldi** [mpv-player/mpv#9014]
### Important Information

Provide following Information:
- mpv version: 0.33.1
- Platform and Version: Currently I'm on Mac OS X but this happens on Windows and Linux as well
- Source of the mpv binary: I installed mpv using Homebrew

### Reproduction steps

Just run this:
`mpv --no-config --player-operation-mode=pseudo-gui --script-opts=osc-visibility=always http://vps.radioduepuntozero.it:9080`

### Expected behavior
The memory usage stays quite the same during playback

### Actual behavior
The memory occupation starts growing, at first really fast (like 2MB per second) then it progressively slows down, but still growing. I've seen it up to more than 600MB after a few days of playback.
It seems that some events actually drop the RAM usage, like on Mac it drops if I put another maximized window on top of the MPV one and then I focus back on the MPV windows. I did this before quitting MPV, in the log I attach.

### Log file

https://0x0.st/-O73.txt

**Comment by CounterPillow** [comment]
my guess is that it's the libass glyph cache.

**Comment by ale-rinaldi** [comment]
> my guess is that it's the libass glyph cache.

@CounterPillow is libass used also to render the OSD graphic itself? Because this is an audio stream with no subtitles.

**Comment by CounterPillow** [comment]
Yes, the entire pseudo-gui of mpv is rendered with ASS subtitles.

**Comment by ale-rinaldi** [comment]
> Yes, the entire pseudo-gui of mpv is rendered with ASS subtitles.

Got it, thanks for clarifying. Unfortunately I have little to no experience with C so it would take me ages to look into it :(

If someone is willing to help me I'd be really grateful, this is a major issue for me since I'd need to keep it open h24 and the RAM will eventually saturate.

**Comment by CounterPillow** [comment]
the libass glyph cache is of limited size, so it's not going to exhaust all of your RAM.

**Comment by tantei3** [comment]
It doesn't seem to be related to libass, even if you disable the pseudo-gui, RAM usage will keep increasing. This seems to be because the stream data is being integrated and cached in mpv, which allows seeking backwards in the stream,

**Comment by avih** [comment]
> run this:
> `mpv --no-config --player-operation-mode=pseudo-gui --script-opts=osc-visibility=always http://vps.radioduepuntozero.it:9080`

> The memory occupation starts growing, at first really fast (like 2MB per second) then it progressively slows down, but still growing. I've seen it up to more than 600MB after a few days of playback.

I don't doubt the facts (memory growth), but the info we have so far doesn't mean it's related to `osc-visibility=always`.

To reach this conclusion, you'd have to also try with `osc-visibility=never` (or just `--no-osc` to not even load the osc script) and confirm that the memory growth issue goes away.

Have you done that?

**Comment by ale-rinaldi** [comment]
Hello,

actually, the RAM usage progressively increases also without the OSD, but also with no GUI at all, using MPV in the terminal.

However, the increase is slower and never goes over 250MB or so. It's still pretty much, but acceptable.

With the OSD on, it increases much faster. On one of the machines, where I have mpv running since about 3 days, it's now at 697MB. It could be me, but it seems it has an increase peak whenever the title changes.

I'm pretty sure the RAM increase only slows down but never really stop, since I used to run it on a Raspberry with 2 GB of RAM and it eventually saturated after 2 or 3 weeks.

**Comment by avih** [comment]
Sure, with OSD/OSC it uses more ram (and grows quicker initially), because libass has a big cache. However, that should eventually stop growing. I don't know how big is the libass cache, but I think it's several hundreds megabytes.

So really, if you claim the issue is due to the OSC, then you need to provide more concrete evidence. For example, two instances of mpv playing the same source/stream, one with OSC and one without OSC, and the one with OSC keeps using more and more memory forever, while the one without stops growing after some time.

Leak bugs are of course possible, but so far you did not provide enough evidence to suggest either that it's related to the OSC, or that there's a leak at all.

**Comment by jaseg** [comment]
Over at https://github.com/jaseg/python-mpv/issues/233 someone observed a simililar issue. What I noticed there is that on the following stream: https://media-ice.musicradio.com/CapitalMP3 mpv still seems to grow by a few hundred kB every other minute even with --demuxer-max-bytes=100k --demuxer-max-back-bytes=100k. It seems that that might be in part due to mpv's processing of the stream's Icy metadata, since it seemed to slow down (but did not stop) when also passing `--http-header-fields='Icy-MetaData: 0'` (which the stream server seemed to respect, since mpv did no longer output the stream title metadata).

**Comment by kasper93** [comment]
The memory increase you are seeing are cumming from libass caches. The size that mpv uses is "reasonable default" from libass. See https://github.com/libass/libass/blob/695509365f152bd28720a0c0e036d46836ee9345/libass/ass_render.h#L43-L47

If someone is interested those sizes can be exposed as mpv option to control how much memory can be used. Maybe @llyyr would like to add it, as he is doing stuff around libass in general.

Eventually it achieve steady state. What you can do to workaround this is to disable graphs in stats.lua, which are more a gimmic than useful things and are slow to render and causing each update to be new cache. See the options of stats.lua how to disable the graphs.

Here is memory usage analysis of stats.lua page 2 updating constantly for 5 minutes. Sorry, for presentation, but there is nothing interesting to see anyway, just a bunch of cached things. Nothing is leaking and it approaches the limit.

First graph is total memory usage, the rest is top allocations by size.
![Image](https://github.com/user-attachments/assets/79c612f6-4b7c-4882-9317-b7451477f575)

**Comment by llyyr** [comment]
> During the viewing of a 2-minute video, the RAM leak was about 200 MB, with mpv shader stats and uosc interface elements (timeline) active, during the viewing of 17 minutes of an hour and a half long video, with uosc interface elements (timeline) active, it was about 400 MB.

I can of course introduce an option to set prune delay for the mpv OSD track, however this sounds like a memory leak somewhere. The RAM uses should not grow indefinitely, and it should stop once it hits the libass cache limit. Pruning it is only hiding the problem, if it is indeed a leak or cache mismanagement somewhere.

**Comment by verygoodlee** [comment]
> What you can do to workaround this is to disable graphs in stats.lua, which are more a gimmic than useful things and are slow to render and causing each update to be new cache. See the options of stats.lua how to disable the graphs.

Currentlly, there're no options to disable the graphs on page 3  #13894,

**Comment by kasper93** [comment]
@llyyr I think you should be using `ass_set_cache_limits()` instead of prune.

**Comment by llyyr** [comment]
Changed the PR to add osd-glyph-limit and osd-bitmap-max-bytes. You can see if that helps

**Comment by llyyr** [comment]
> Since I do not know, if there are any edge cases where libass needs its caches, I think it is safer to use values like (it is not necessarily these values, you can adjust them from your experience): osd-prune-delay=15, osd-glyph-limit=512, osd-bitmap-max-size=32.

The PR default is effectively no pruning, and glyph limit 10000 and bitmap cache 128 MB. I don't think there's any reason to change them, but the option is there if the user wants to. You also don't need to have stats.lua open to watch videos, it's probably better to not watch videos with stats.lua shown to have visual clarity and less obstruction.

**Comment by kasper93** [comment]
I've tested with the new options added in #16226. All libass caching is under control after settings options. The remaining memory increase comes from harfbuzz cache, which is itself caching things, but it is not as much as with libass caches. We don't have direct control of harfbuzz caches form mpv side. It seems to cache 12MB in one cache and ~2MB in few others, but nothing excessive.

I don't see any more issues to fix here.

> So, as a result, for now, I configured uosc not to show timeline in windowed mode.

If you set `osd-glyph-limit=1` and `osd-bitmap-max-size=1` there shouldn't be any persistent memory usage grow.
